addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/TestInfoSessionIdURL.java

```diff
@@ -132,25 +135,36 @@ public void setParent(PassiveScanThread parent) {
     }
 
     /**
-     * Scan the request. Currently it does nothing.
+     * Scan the response. Currently it does nothing. TODO: This method should really
+     * scan the contents of the response to see if it is HTML, and if so, look for
+     * HREFs and SRC tags and check if the URLs in them contain session IDs. This
+     * would enable ZAP to detect Session IDs in URLs exactly where they are
+     * occurring, rather than simply detecting the symptom of the problem in URLs of
+     * requests.
      *
-     * @param msg the HTTP message
-     * @param id the id of the request
+     * @param msg    the HTTP message
+     * @param id     the id of the response
+     * @param source the source code of the response
      */
     @Override
-    public void scanHttpRequestSend(HttpMessage msg, int id) {
-        // do Nothing it's related to response managed
+    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+        // do Nothing. All work currently is done scanning the request.
     }
 
+    // Pattern used only in scanHttpRequestSend() method below.
+    private static final Pattern PATHSESSIONIDPATTERN =
+            Pattern.compile(
+                    "(jsession|cf)id=[\\dA-Z]{" + SESSION_TOKEN_MIN_LENGTH + ",}",
+                    Pattern.CASE_INSENSITIVE);
+
     /**
-     * Perform the passive scanning of URL based session IDs
+     * Perform passive scanning for URL based session IDs in the HTTP request
      *
-     * @param msg the message that need to be checked
+     * @param msg the message that needs to be checked
      * @param id the id of the session
-     * @param source the source code of the response
      */
     @Override
-    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
```

Comment:
This should be reverted, the response is being scanned.

Reply:
I disagree, the response is only being scanned if a session ID is found in the request. The bulk of the scanning is on the request, not the response. That's why this change was made.

Reply:
That means it's being scanned so it should be in this method (which is where the response is guaranteed to exist). If the response was not being checked I'd agree, it would make more sense in the other method.

Reply:
OK, but the counterargument is that the request IS being scanned, and that method would then be empty if we switch back, and that's confusing. Given that it STARTS with scanning the request, and only looks at the response under special conditions, I think this change should be kept as is.  Your argument is that if it IS being scanned, it should go in that method. Well, the request IS being scanned, so this code belongs in this method ... :-)

Reply:
Agree, it's not a clear case and we can have it both ways. I'm fine leaving as is only if `checkSessionIDExposureTo3rdParty` is changed to return immediatelly if `msg.isResponseFromTargetHost()` is `false` (that addresses my concern about whether the (real) response exists or not).

Reply:
It’s only scanning the request in that the URI is solely an element of the request, it’s not parsing request headers or body. 

If operating on or retrieving URI automatically implicated the scan request method then all the passive scanners need to be rewritten to retrieve and globally store the URI in the scan request method so that it can be used in the scan response method for raising alerts.

Reply:
I have no idea what you are talking about.  Its not the scan**All**HttpRequestSend method. What does scanning ALL of the request (or not) have to do with anything. Intuitively, the scanner is scanning the REQUEST for issues, and it only looks at the response to get additional details if it finds a problem in the REQUEST.  I'm not going to fall on my sword over this. Its your app and if you really want to move it back, we will, but this is VERY confusing to us trying to maintain/improve this code.

Reply:
Sorry to confuse you further. 

> but this is VERY confusing to us trying to maintain/improve this code.

That's fair, I remember that curve when I encountered it 5 (maybe more?) years ago. My suggestion is stick to the specific issue/task at hand. If there are other things that occur to you, tackle them separately or after discussion.
Unfortunately with Open Source Software you're basically always working on or with something that someone else built, which means likely very differently from how you'd do or expect things. It's an adaptation test :smiley: (JK)

Anyway there's no need to debate this further, @thc202's latest response provides a good way forward :+1: